server/remote_cache/chunked_manifest/chunking.go

```diff
@@ -0,0 +1,197 @@
+	for i, outputFile := range ar.GetOutputFiles() {
+		indexStr := strings.TrimPrefix(outputFile.GetPath(), chunkedManifestPrefix+"chunk_")
+		expectedIndex := i
+		actualIndex, err := strconv.Atoi(indexStr)
+		if err != nil || actualIndex != expectedIndex {
+			return nil, status.InvalidArgumentErrorf(
+				"invalid chunked manifest: expected %q but got path %q",
+				chunkedManifestPrefix+"chunk_"+strconv.Itoa(expectedIndex),
+				outputFile.GetPath())
+		}
+
+		chunkDigests[i] = outputFile.GetDigest()
+	}
+
+	return &ChunkedManifest{
+		BlobDigest:     blobDigest,
+		ChunkDigests:   chunkDigests,
+		InstanceName:   instanceName,
+		DigestFunction: digestFunction,
+	}, nil
+}
+
+// Store saves the chunked manifest to the Action Cache using the blob digest as
+// the AC key.
+func (cm *ChunkedManifest) Store(ctx context.Context, cache interfaces.Cache) error {
+	log.Debugf("Storing chunked manifest: blob=%s size=%d chunks=%d instance=%s",
+		cm.BlobDigest.GetHash(), cm.BlobDigest.GetSizeBytes(), len(cm.ChunkDigests), cm.InstanceName)
+
+	ar, err := cm.encodeToActionResult()
+	if err != nil {
+		return err
+	}
+
+	arBytes, err := proto.Marshal(ar)
+	if err != nil {
+		return status.InternalErrorf("marshal chunked manifest to ActionResult: %w", err)
+	}
+
+	acInstanceName := chunkedManifestPrefix + cm.InstanceName
+	acRN := digest.NewResourceName(cm.BlobDigest, acInstanceName, rspb.CacheType_AC, cm.DigestFunction)
+	if err := acRN.Validate(); err != nil {
+		return err
+	}
+
+	log.Debugf("Storing chunked manifest with AC key: blob=%s/%d digest_func=%s instance=%q ac_key=%s",
+		cm.BlobDigest.GetHash(), cm.BlobDigest.GetSizeBytes(), cm.DigestFunction, acInstanceName, acRN.ToProto().String())
+
+	if err := cache.Set(ctx, acRN.ToProto(), arBytes); err != nil {
+		return status.InternalErrorf("store chunked manifest in AC: %w", err)
+	}
+
+	log.Debugf("Stored chunked manifest in AC: blob=%s manifest_size=%d", cm.BlobDigest.GetHash(), len(arBytes))
+	return nil
+}
+
+// Load retrieves a chunked manifest from the Action Cache.
+func Load(ctx context.Context, cache interfaces.Cache, blobDigest *repb.Digest, instanceName string, digestFunction repb.DigestFunction_Value) (*ChunkedManifest, error) {
+	chunkedManifestInstanceName := chunkedManifestPrefix + instanceName
+	acRN := digest.NewResourceName(blobDigest, chunkedManifestInstanceName, rspb.CacheType_AC, digestFunction)
+	if err := acRN.Validate(); err != nil {
+		return nil, err
+	}
+
+	log.Debugf("Looking up chunked manifest: blob=%s/%d digest_func=%s instance=%q ac_key=%s",
+		blobDigest.GetHash(), blobDigest.GetSizeBytes(), digestFunction, chunkedManifestInstanceName, acRN.ToProto().String())
+
+	arBytes, err := cache.Get(ctx, acRN.ToProto())
+	if err != nil {
+		if status.IsNotFoundError(err) {
+			return nil, status.NotFoundErrorf("chunked manifest not found for blob %s", blobDigest.GetHash())
+		}
+		return nil, status.InternalErrorf("retrieve chunked manifest from AC: %w", err)
+	}
```

Comment:
Worth noting that you are getting the AC entry directly from the lower-level Cache interface here, which won't validate if all the output files are still available.

The logic for that lives here https://github.com/buildbuddy-io/buildbuddy/blob/ae198522df4d6e5248be33966e4a0c963c9d4c83/server/remote_cache/action_cache_server/action_cache_server.go#L79 so I suspect you should be using `action_cache_server` instead of the Cache interface directly?